src/server/game/AI/SmartScripts/SmartAI.cpp

```diff
@@ -650,9 +650,7 @@ void SmartAI::OnCharmed(bool apply)
 {
     GetScript()->ProcessEventsFor(SMART_EVENT_CHARMED, NULL, 0, 0, apply);
 
-    if (!apply && !me->IsInEvadeMode() && me->GetCharmerGUID())
-        if (Unit* charmer = ObjectAccessor::GetUnit(*me, me->GetCharmerGUID()))
-            AttackStart(charmer);
+    CreatureAI::OnCharmed(apply);
```

Comment:
I debated this with @Shauren some time ago, and this is not ok.

He claimed that AI change should not be done (it will prevent smartscripts to trigger)

I dont know the right solution, but this is what i came up with:

* Enable AI change only on CHARM_TYPE_CHARM
This is like a middle term solution, but needs also to check for 
- If creature is vehicle
- if creature is possessed
If charm is one of those cases just store the ReactState to fix the aggro behaviour
Else use the ````CreatureAI::OnCharmed(apply)```` to change into the PetAI

* Enable the AI change just as u did
Prevents any smartscript to trigger, and there are some creatures that use them while charmed
(texts for example)

* Implement PetAI functionallities inside the SmartAI
That would be the perfect solution but a huuuuuuuuuge load of work:
- Autocast
- Follow
- Disable events with cast action (or just disable cast without a special flag)
- ...